Convert MdiSubWindow to use dynamic factory

samueljackson92 opened the pull request:
This converts classes derived from MdiSubWindow which are serialised to Mantid project files to be created by a new dynamic factory, drastically simplifying how each object is loaded at a high level. 

Each window type can now be created using a default constructor from the factory. The ```loadFromProject``` function will then initialise and load setting for each window. The loading logic for each type of window is now contained within the class itself, i.e. MdiSubWindows now completely know how to load themselves from file. 

Unfortunately there is still a large dependance on ApplicationWindow, but reducing that would be getting beyond the scope of this ticket.

This PR should not add any new functionality or fix any bugs, but hopefully should not break anything either.

**To test:**

A lot of code for project loading/saving has been moved around, but on the surface it should appears as if nothing has changed.

 - Check all the tests pass. (especially MantidPlotProjectSerialise!)
 - Check that all types of windows can be saved and loaded back. For a list of what should and shouldn't work see issue #15602. Types of windows should include:
   - MantidMatrix
   - Matrix
   - Note
   - Table
   - TableStatistics
   - 1D plots
   - 2D plots
   - 3D plots
   - InstrumentWindow
   - ScriptWindow

- Code review.

Fixes #16768.

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

samueljackson92 commented:
Talking to Martyn, we should keep the default contractors and switch to using a [variadic function templates](http://en.cppreference.com/w/cpp/language/parameter_pack) to create a factory that allows construction of objects with non-default constructors. 

Unfortunately the dynamic factory implementation does not support this already and modifying it to achieve this would:
 1. Not work because the ```createInstance``` method(s) on the factory are virtual and cannot be templated.
 2. Touch too many things beyond the scope of this ticket.

Therefore after much discussion we'll proceed by creating a new, light weight dynamic factory implementation that functions similarly to the standard dynamic factory.

AntonPiccardoSelg commented on MantidPlot/src/Graph3D.cpp:
```diff
@@ -2483,120 +2494,437 @@ Graph3D::~Graph3D() {
+  if (valVec.size() < 2)
+    return nullptr;
+
+  QString caption = QString::fromStdString(valVec[0]);
```
you can use `auto` here and above/below

AntonPiccardoSelg commented on MantidPlot/src/Graph3D.h:
```diff
@@ -77,9 +81,43 @@ class Graph3D : public MdiSubWindow, public Mantid::IProjectSerialisable {
     Cones = 4
   };
 
+  enum SurfaceFunctionType {
```
is an enum class an option here?

samueljackson92 replied:
@AntonPiccardoSelg Yes. My C++ language knowledge grows daily.

AntonPiccardoSelg commented on MantidPlot/src/WindowFactory.h:
```diff
@@ -0,0 +1,203 @@
+#endif /* _WIN32 */
+
+typedef Mantid::Kernel::SingletonHolder<WindowFactoryImpl> WindowFactory;
+}
```
This has changed recently: https://github.com/mantidproject/mantid/wiki/Exporting-Symbols-of-Shared-Libraries-in-Mantid

samueljackson92 replied:
As discussed this isn't applicable because it's in MantidPlot.

martyngigg replied:
You're correct that the symbol exporting doesn't apply to MantidPlot but this code suggests, with the comments etc that it is involved in that mechanism somehow.

I think we should clean up the code at the bottom here so that it more closely follows that as defined in the link provided by @AntonPiccardoSelg. At the very least the `SingletonHolder` template needs to be specialized in the Mantid::Kernel namespace. I think all we need to do is remove the `EXTERN_MANTID_API` from the example.

AntonPiccardoSelg commented on MantidPlot/src/WindowFactory.h:
```diff
@@ -0,0 +1,203 @@
+    return _stricmp(s1.c_str(), s2.c_str()) < 0;
+#else
+    return strcasecmp(s1.c_str(), s2.c_str()) < 0;
+#endif
```
* Can this be reused from above (or the other way around)?
* Can you use the `boost::iequals`?

samueljackson92 replied:
@AntonPiccardoSelg Yes. This code is also in ```DynamicFactory.h``` which would be worth replacing as well (although perhaps not in this issue). Is there a sensible common place we can put this comparator?

AntonPiccardoSelg replied:
Maybe it is worth checking with the person who wrote this in the first place, why they did it. It could well be that it is much more performant, although chances are that the boost version does the same thing just abstracted away.

I am not aware of a sensible place, but you could create one. If DynamicFactory uses the same thing it would have to be in `Kernel`, but it might be worth getting Martyn's opinion on this.

AntonPiccardoSelg commented:
There are quite a few opportunities for the usage of `auto`. If they were added during this work, it might be a good opportunity to change that, if they were there before, then don't worry too much about it.

samueljackson92 commented:
@AntonPiccardoSelg a lot of this is code that has simply been moved around. Future tickets will be bug fixes & improvements now loading code is in the same place, so it might be nicer to catch it then. However I have replaced some which I was aware that I had added (e.g. the Graph3D* you pointed out above).

AntonPiccardoSelg commented:
@samueljackson92 Yes, then don't worry about it.

samueljackson92 commented on MantidPlot/src/TSVSerialiser.h:
```diff
@@ -33,6 +33,17 @@
+    return stricmp(s1.c_str(), s2.c_str()) < 0;
+#else
+    return strcasecmp(s1.c_str(), s2.c_str()) < 0;
+#endif
```
I've changed my mind on this one:
 1. Note that be above is not equality, but checking if s1 is less than s2. 
 2. Boost versions add locale support, which we are not bothered about and the C lib ones should therefore be more efficient in this case.
 3. Yes this isn't the nicest code, but after the next few commits there should be only one instance.

What do you think? If you don't agree I could find the equivalent compare function for boost?

AntonPiccardoSelg commented:
@samueljackson92 There seems to be a merge conflict.

AntonPiccardoSelg commented:
*Code look good to me.
*Saved all types of windows in a project and reloaded them. Worked fine on Windows and Ubuntu. 
Happy to :shipit:

martyngigg commented on Framework/Kernel/inc/MantidKernel/case_insensitive_map.h:
```diff
@@ -0,0 +1,31 @@
+#ifndef CASE_INSENSITIVE_MAP_H
```
Is there a reason that the filename doesn't follow the standard conventions without underscores?

samueljackson92 replied:
Not really. Wasn't sure if we used camel case for just classes and lower case underscore for stuff that isn't e.g. ```make_*.h``` or whether that was just due mimicking std lib support.

martyngigg replied:
I think given that the alias class name is CamelCase then it would be better if the filename reflected that. Could you change it to `CaseInsensitiveMap`?

samueljackson92 replied:
Yes, seems like a silly decision in hindsight. Already made the change, just checking it builds.

martyngigg commented:
Thanks for making that update. Happy to merge now.
